src/main/java/org/terracotta/offheapstore/paging/UpfrontAllocatingPageSource.java

```diff
@@ -440,67 +473,146 @@ public synchronized Runnable removeAllocationThreshold(ThresholdDirection direct
+
+      final PrintStream allocatorLog = createAllocatorLog(toAllocate, maxChunk, minChunk);
 
-      PrintStream allocatorLog;
       try {
-        allocatorLog = createAllocatorLog(max, maxChunk, minChunk);
-      } catch (IOException e) {
-        LOGGER.warn("Exception creating allocation log", e);
-        allocatorLog = null;
-      }
-      long start = System.nanoTime();
-      if (allocatorLog != null) {
-        allocatorLog.printf("timestamp,duration,size,allocated,physfree,totalswap,freeswap,committed%n");
-      }
-      long progressStep = Math.max(PROGRESS_LOGGING_THRESHOLD, (long) (max * PROGRESS_LOGGING_STEP_SIZE));
-      long lastFailureAt = 0;
-      long currentChunkSize = maxChunk;
-      long allocated = 0;
-      long nextProgressLogAt = progressStep;
-      while (allocated < max) {
-        long blockStart = System.nanoTime();
-        ByteBuffer b = source.allocateBuffer((int) Math.min(currentChunkSize, (max - allocated)));
-        long blockDuration = System.nanoTime() - blockStart;
-        if (b == null) {
-          if (fixed || (currentChunkSize >>> 1) < minChunk) {
-            throw new IllegalArgumentException("An attempt was made to allocate more off-heap memory than the JVM can allow." +
-                    " The limit on off-heap memory size is given by the -XX:MaxDirectMemorySize command (or equivalent).");
-          } else {
-            LOGGER.debug("Allocated {}B in {}B chunks.", new Object[]{DebuggingUtils.toBase2SuffixedString(allocated - lastFailureAt), DebuggingUtils.toBase2SuffixedString(currentChunkSize)});
-            currentChunkSize >>>= 1;
-            lastFailureAt = allocated;
-          }
-        } else {
-          buffers.add(b);
-          allocated += b.capacity();
-          
-          if (allocatorLog != null) {
-            allocatorLog.printf("%d,%d,%d,%d,%d,%d,%d,%d%n", System.nanoTime() - start, blockDuration, b.capacity(), allocated, PhysicalMemory.freePhysicalMemory(), PhysicalMemory.totalSwapSpace(), PhysicalMemory.freeSwapSpace(), PhysicalMemory.ourCommittedVirtualMemory());
-          }
-          if (allocated > nextProgressLogAt) {
-            LOGGER.info("Allocation {}% complete", (100 * allocated) / max);
-            nextProgressLogAt += progressStep;
+        if (allocatorLog != null) {
+          allocatorLog.printf("timestamp,duration,size,physfree,totalswap,freeswap,committed%n");
+        }
+
+        ExecutorService executorService = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors());
+
+        List<Future<Long>> futures = new ArrayList<Future<Long>>((int)toAllocate / maxChunk + 1);
+
+        for (long dispatched = 0; dispatched < toAllocate; ) {
+          final int currentChunkSize = (int)Math.min(maxChunk, toAllocate - dispatched);
+          futures.add(executorService.submit(new Callable<Long>() {
+            @Override
+            public Long call() throws Exception {
```

Comment:
Wouldn't it be more natural for the `Callable` to return the `ByteBuffer` it allocated? That way we could avoid having to pass in the collection here.